### PR TITLE
Add `x86_64-pc-cygwin` target to `advanced.yml`

### PR DESCRIPTION
**Pull request by Blackhex**
Adds Cygwin and `x86_64-pc-cygwin` GCC build to the `advanced.yml` workflow.

### CODE REVIEW

File: .github/workflows/advanced.yml [hunk 1]
```diff
@@ -4,7 +4,6 @@ on:
   push:
     branches:
       - main
-  pull_request:
```
Comment (Blackhex) [inline .github/workflows/advanced.yml:7]: To be removed.

File: .github/workflows/cygwin.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+name: Cygwin build
+
+on:
+  pull_request:
```
Comment (Blackhex) [inline .github/workflows/cygwin.yml:4]: To be removed before merge.

File: .github/scripts/config.sh [hunk 3]
```diff
@@ -53,6 +53,8 @@ TESTS_PACKAGE_NAME=${TESTS_PACKAGE_NAME:-$TOOLCHAIN_NAME-tests.tar.gz}
 FFMPEG_PATH=${FFMPEG_PATH:-~/ffmpeg}
 FFMPEG_TESTS_PATH=${FFMPEG_TESTS_PATH:-~/ffmpeg-tests}
 
+CYGWIN_PATH=${CYGWIN_PATH:-~/cygwin}
```
Comment (eukarpov) [inline .github/scripts/config.sh:56]: maybe we need to specify target in cygwin name.
I guess cygwin for x64 should be built as validation.

File: .github/scripts/cygwin/build.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+source `dirname ${BASH_SOURCE[0]}`/../config.sh
+
+CYGWIN_BUILD_PATH=$BUILD_PATH/cygwin
```
Comment (eukarpov) [inline .github/scripts/cygwin/build.sh:5]: the same idea for build target

File: .github/scripts/cygwin/build.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+            --target=$TARGET
+    echo "::endgroup::"
+fi
+
+echo "::group::Build Cygwin"
+    make $BUILD_MAKE_OPTIONS
+echo "::endgroup::"
+
+if [ "$RUN_INSTALL" = 1 ] ; then
+    echo "::group::Install OpenSSL"
```
Comment (eukarpov) [inline .github/scripts/cygwin/build.sh:25]: OpenSSL? :)

File: .github/scripts/download-toolchain.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+source `dirname ${BASH_SOURCE[0]}`/config.sh
+
+echo "::group::Download toolchain"
+    RELEASE_NAME=`gh release view --json name --jq .name`
+    rm -f $ARTIFACT_PATH/$TOOLCHAIN_PACKAGE_NAME
```
Comment (eukarpov) [inline .github/scripts/download-toolchain.sh:7]: what if $RELEASE_NAME is empty?

File: .github/workflows/cygwin.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
+name: Build cygwin
+
+on:
+  pull_request:
+  workflow_dispatch:
+    inputs:
+      cygwin_branch:
+        description: 'Cygwin branch'
+        required: true
```
Comment (eukarpov) [inline .github/workflows/cygwin.yml:9]: could be false with default branch name?

File: .github/workflows/cygwin.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        default: 'main'
+  workflow_call:
+    inputs:
+      cygwin_branch:
+        type: string
```
Comment (eukarpov) [inline .github/workflows/cygwin.yml:14]: what about using the same default branch?

File: .github/scripts/config.sh (hunk 3 above)
Comment (Blackhex) [reply .github/scripts/config.sh:56]: Good point.

File: .github/scripts/cygwin/build.sh (hunk 4 above)
Comment (Blackhex) [reply .github/scripts/cygwin/build.sh:5]: Here it's not relevant as the `$BUILD_PATH` already contains target string.

File: .github/scripts/download-toolchain.sh (hunk 6 above)
Comment (Blackhex) [reply .github/scripts/download-toolchain.sh:7]: I am not sure I understand the merrit of the question. The relase name will be empty only when there is no release. In that case the download fails witch is IMO the expected behavior.

File: .github/workflows/cygwin.yml (hunk 8 above)
Comment (Blackhex) [reply .github/workflows/cygwin.yml:14]: Added.

File: .github/scripts/download-toolchain.sh (hunk 6 above)
Comment (eukarpov) [reply .github/scripts/download-toolchain.sh:7]: maybe it is better to check if some release has been found

**Review by eukarpov — APPROVED** [review]

**Comment by Blackhex** [comment]
This PR is completely wrong: One cannot build Cygwin without stage 1 `???-w64-cygwin` GCC and  stage 2 `???-w64-cygwin` GCC cannot be build without Cygwin. Closing this PR and I'll create another one that will add the whole build process to the `advanced.yml` as another matrix platform.

**Comment by Blackhex** [comment]
Completely re-done and ready for review.

File: .github/scripts/build.sh [hunk 9]
```diff
@@ -7,29 +7,50 @@ if [ "$RUN_BOOTSTRAP" = 1 ]; then
...
+fi
+if [[ "$PLATFORM" =~ mingw ]]; then
     .github/scripts/toolchain/build-mingw-winpthreads.sh
 fi
+if [[ "$PLATFORM" =~ cygwin ]]; then
+    .github/scripts/toolchain/build-cocom.sh
+    .github/scripts/toolchain/build-cygwin-crt.sh
+fi
+
+if [[ "$PLATFORM" =~ cygwin ]]; then
```
Comment (eukarpov) [inline .github/scripts/build.sh:42]: Should it be joined with previous condition?

File: .github/scripts/toolchain/build-cygwin.sh [hunk 10]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+
+source `dirname ${BASH_SOURCE[0]}`/../config.sh
```
Comment (eukarpov) [inline .github/scripts/toolchain/build-cygwin.sh:3]: I think build-cygwin.sh and build-cygwin-crt.sh could be united and configured slightly different.

**Review by eukarpov — APPROVED** [review]

**Pull request merged**